File: vigridr/src/games/tictactoe/sampleclient/ClientLogic.py
from WorldModel.ttypes import Marker
from Command.ttypes import Command
from Command.ttypes import Coordinate
from GameModel.ttypes import GameStatus
from GameResult.ttypes import GameResult

from random import randint

class Solution:
    def __init__(self, gameInit):
        """
        Constructor: called at the beginning of the game.
        You may do initialization here.

        Parameter:
            gameInit - contains a field named gameDescription, which itself contains a field named myType.
                       myType is of type int, and can have one of the Marker values: UNMARKED, X and O.
        """
        print "Python Client"
        print "PlayerType: " + Marker._VALUES_TO_NAMES[gameInit.gameDescription.myType]

    def play_turn(self, wm, turn):
        """
        This method is called once for every turn.
        This specific example solution returns a random valid position.

        Parameters:
            wm   - an instance of the WorldModel class that contains a field called table which is a list of lists of ints,
                   which can have on of the Marker values.
            turn - an integer, the index of the turn.
                   If you receive twice the same number, don't worry, just ignore it.

        Returns:
            A Command instance - a Command contains a field called coordinate of class coordinate.
                                 A Coordinate contains two fields of type int, x and y.
        """
        command = Command(Coordinate())
        while True:
            x = randint(0,2)
            y = randint(0,2)
            if wm.table[x][y] == Marker.UNMARKED:
                command.coordinate.x = x
                command.coordinate.y = y
                break
        print "%d: %s" % (turn, repr(command.coordinate))

        return command

    def end_of_game(self, result):
        """
        This method is called at the end of the game.

        Parameters:
            result - an integer, which can be GameResult.WON, GameResult.TIED or GameResult.LOST.
        """
        print "End of game - " + GameResult._VALUES_TO_NAMES[result]

File: vigridr/src/games/tictactoe/sampleclient/ClientLogic.cpp
#include "ClientLogic.h"

#include <stdlib.h>
#include <time.h>
#include <iostream>

using ::mjollnir::vigridr::Command;
using ::mjollnir::vigridr::Coordinate;
using ::mjollnir::vigridr::WorldModel;
using ::mjollnir::vigridr::GameInit;
using ::mjollnir::vigridr::Marker;

/*
 * A helper function to print a Coordinate on screen.
 */
std::ostream& operator<<(std::ostream& os, Coordinate c) {
    return os << "Coordinate(" << c.x << ", " << c.y << ")";
}

/*
 * This function is called at the beginning of the game.
 * You may do initialization here.
 *
 * Parameter:
 *     gameInit - contains a field named gameDescription, which itself contains a field named myType.
 *                myType is of type Marker, which is an enum. Marker has three fields: UNMARKED, X and O.
 */
void init(const GameInit& gameInit) {
  srand(time(NULL));

  std::cout << "C++ Client" << std::endl;
  std::cout << "PlayerType: " << (gameInit.gameDescription.myType == Marker::X ? "X" : "O") << std::endl;
}

/*
 * This function is called once for every turn.
 * This specific example solution returns a random valid position.
 *
 * Parameter:
 *     wm - an instance of the WorldModel class that contains a field called table which is a vector of vectors of Markers.
 *
 * Returns:
 *     A Command instance - a Command contains a field called coordinate of class Coordinate.
 *                          A Coordinate contains two fields of type int32_t, x and y.
 */
Command playTurn(const WorldModel& wm) {
  Command command;

  while(true) {
    size_t x = rand()%3;
    size_t y = rand()%3;
    if (wm.table[x][y] == Marker::UNMARKED) {
      command.coordinate.x = x;
      command.coordinate.y = y;
      break;
    }
  }

  std::cout << command.coordinate << std::endl;
  return command;
}

